### ISSUE TITLE
Adjust for change of Pandoc data model in 1.17.3

### ISSUE THREAD

**Issue by nichtich** [nichtich/Pandoc-Elements#49]
In pandoc 1.17.3, the JSON representation changes quite a bit:
- The toplevel JSON format is now `{"pandoc-api-version" : [MAJ, MIN, REV], "meta" : META, "blocks": BLOCKS}` instead of `[{"unMeta": META}, [BLOCKS]]`
- Leaf nodes no longer have an empty array for their "c" value. Thus, for example, a `Space` is encoded as `{"t":"Space"}` rather than `{"t":"Space","c":[]}` as before.

**Comment by bpj** [comment]
Have you seen the announcement? 

https://groups.google.com/d/msgid/pandoc-discuss/20161026224653.GA40381%40Johns-MBP.home

I can start working on this right away as I'm free at least until Monday.

**Comment by nichtich** [comment]
> I can start working on this right away as I'm free at least until Monday.

Yes that would be great!

Jakob

**Comment by nichtich** [comment]
I added Pandoc::Version in https://github.com/nichtich/Pandoc-Wrapper (not released yet) to be used by this extension. Any opinions on that?

**Comment by bpj** [comment]
Just beware that Test::More::is_deeply() botches up on stringified objects, comparing their stringification rather than their structure. [Test::Deep](https://metacpan.org/pod/Test::Deep) has a fix. See also [my modified test file](https://github.com/nichtich/Pandoc-Elements/pull/51/files#diff-33be3709b09e5cd79904988b71fd9b87).

**Comment by nichtich** [comment]
Good to know. I think this only affects `t/parse.t` and the test is skipped for pandoc >= 1.18. I plan to release a new Pandoc(-Wrapper) module and we can adjust Pandoc-Elements afterwards.

**Comment by bpj** [comment]
It affects `Pandoc-Elements/t/synopsis.t`, which my fix addresses. The other tests went OK when I had an overloaded object.

**Comment by ambs** [comment]
A filter written by @bpj is not working. I think it is given that change (??).
A quick fix would be very welcome.

**Comment by nichtich** [comment]
@ambs can you give an example please (which filter?) We are working on support of Pandoc 1.18.

**Comment by ambs** [comment]
https://gist.github.com/ambs/06348379bf4b183194aaa99ad64096e5

**Comment by ambs** [comment]
ok, line 29 of Pandoc::Filter:

`````
    my $ast    = Pandoc::Elements::pandoc_json(<STDIN>);
``````
two lines below, `$ast->content` is called, but `$ast` is not a blessed ref. This is `$ast` conents:

`````
$VAR1 = {
          'pandoc-api-version' => [
                                    1,
                                    17,
                                    0,
                                    4
                                  ],
          'meta' => {
                      'author' => {
                                    't' => 'MetaList',
                                    'c' => [
                                             bless( {
                                                      'c' => [
...
``````

@bpj, HTH.

**Comment by nichtich** [comment]
@ambs Sorry for not fixing this faster. I think both of @bgj and me are busy. I will come back to this issue at Tuesday or Wednesday.

**Comment by ambs** [comment]
sure! Just poking as sometimes it seems github forgets to notify users :-)

**Comment by bpj** [comment]
I'm working on it ATM, in fact!

**Comment by bpj** [comment]
@ambs, do you mind sharing your document? I still can't reproduce. You can send it to me by email.

**Comment by bpj** [comment]
@ambs, I think I got it. It appears you haven't in fact been running the filter with the Pandoc::Elements from my branch, but with the released one, or whichever other you have installed. 
Please unpack the attached archive in the root directory of my branch and run the test inside it with `prove -mlv`, then run the filter and `.md` resp. `.json` files with pandoc (1.18) as normally from the command line and send me their respective merged stdout and stderr. You'll need to install Test::Script if you haven't already.

[test-html-only.zip](https://github.com/nichtich/Pandoc-Elements/files/580364/test-html-only.zip)

TIA,

/bpj

**Comment by ambs** [comment]
I feel ashamed now. I checked out using the URL github gave me, and I didn't notice it checked out the master branch. Shame, shame, shame.

I am very sorry for the extra work debugging something that was already working.
Thank you very much for your help. This branch works great for me.

**Comment by bpj** [comment]
No worries. GitHub can be weird.I made a PR from my master branch once and didn't get as much as a warning.

**Comment by bpj** [comment]
@ambs, don't feel bad. I did find some real bugs so the time wasn't wasted.

**Comment by bpj** [comment]
@nichtich I have made a branch with all cruft removed, with the changes put manually on top of the current master.
https://github.com/bpj/Pandoc-Elements/tree/bpj-pandoc-1.18-simplified
Note I'm not asking you to pull it! At this point I'm mostly curious why my other branch built for me but not elsewhere, so I would appreciate if you try to build it. It builds fine for me and for Travis CI
https://travis-ci.org/bpj/Pandoc-Elements/jobs/175369607

**Comment by nichtich** [comment]
Looks good, except `t/synopsis.t.old` if renamed to `t/synopsis.t` fails.

**Comment by bpj** [comment]
That's why I renamed it. Its functionality is taken over by `t/synopsis-noclass.t`.

**Comment by ambs** [comment]
Damn, updated from CPAN and lost this again.

**Comment by bpj** [comment]
Check out the branch I mentioned yesterday https://github.com/nichtich/Pandoc-Elements/issues/49#issuecomment-260147597. Some live testing is just what it needs ;-)

**Comment by nichtich** [comment]
@bpj ok, I removed `t/synopsis.t.old`, renamed `t/synopsis-noclass.t` to ` t/synopsis.t` and removed  `PANDOC_API_VERSION` - for now it's not tested or documented and I'd prefer to only have one ENV injection. We may further change `t/synopsis.t` to not test the internal structure (unit testing should not rely on internals) but only the serialized JSON. See branch https://github.com/nichtich/Pandoc-Elements/tree/dev for the current state. Do you add documentation of extended `Document` function or shall I?

**Comment by bpj** [comment]
I'll do it tomorrow.

**Comment by bpj** [comment]
@nichtich here is a documentation patch done on your dev branch.

[bpj-1.18-dev-doc.patch.zip](https://github.com/nichtich/Pandoc-Elements/files/589432/bpj-1.18-dev-doc.patch.zip)

The description of the Document function and related changes is rather thorough, since IME it is better to start out that way and shorten later if appropriate.

I also adjusted the synopsis example, caught a few typos and documented the treatment of line blocks when producing old-style JSON.

**Comment by nichtich** [comment]
@bpj ok, I applied the patch (after some rewording) to the same commit in branch `dev` so code and documentation will be one change. If this is ok, I can merge it into master branch. For the remaining implementation I think that setting `$PANDOC_VERSION` should stay the preferred way to explicitly setting serialization, we need a mapping between pandoc executable version and api version but only internally.

**Comment by bpj** [comment]
Sorry for the delay! It looks good. After all this hassle I'm inclined to just expect that the user is intelligent enough to learn and handle the difference between API version and executable version, however there is no actual difference *at present* between checking for API version >= 1.17.0.4 and executable version >= 1.18, so I think it's OK.

**Comment by nichtich** [comment]
Ok, both `master` and `dev` branch are in sync. Shall I continue to add the missing functionality based on your original pull request or do you extend the code?

To clarify my previous comment: I think setting executable version via `PANDOC_VERSION` is enough. api version can be set via Document constructor or by a mapping table, e.g.

~~~perl
my $version = pandoc_version( api => '1.17.0.4' ); # returns version 1.18
local $Pandoc::Elements::PANDOC_VERSION = $version;
~~~

but I don't see the use case of setting api version via `ENV`.

**Comment by bpj** [comment]
How much more do you want to add?  I think it may be time for a release, perhaps after adding that mapping table. I have come to agree that there is probably no use setting api version via ENV.

**Comment by nichtich** [comment]
What's largely missing are tests and a way to set PANDOC_VERSION via api version number (=mapping table).

**Comment by bpj** [comment]
What is still missing when it comes to tests? Perhaps the tests which I wrote for version comparison can be salvaged.

**Comment by bpj** [comment]
@nichtich, what is still missing? I have some time.

**Comment by nichtich** [comment]
I just released 0.27 with support of Pandoc 1.18. Thanks for all contributions! The change turned out to be a bit more complicated and I discovered some bugs and new requirements (see issue list).
